File: src/moltblock/memory.py
"""Working memory: shared state for one task (drafts, critique, final artifact, verification result)."""

from dataclasses import dataclass, field


@dataclass
class WorkingMemory:
    """In-memory scratchpad for a single run of the agent graph."""

    task: str = ""
    draft: str = ""
    critique: str = ""
    final_candidate: str = ""
    verification_passed: bool = False
    verification_evidence: str = ""
    authoritative_artifact: str = ""
    meta: dict = field(default_factory=dict)
    # Graph runner: node_id -> output (filled by graph execution)
    slots: dict = field(default_factory=dict)
    # Injected from long-term memory for agent context (read-only)
    long_term_context: str = ""

    def set_task(self, task: str) -> None:
        self.task = task

    def set_draft(self, draft: str) -> None:
        self.draft = draft

    def set_critique(self, critique: str) -> None:
        self.critique = critique

    def set_final_candidate(self, candidate: str) -> None:
        self.final_candidate = candidate

    def set_verification(self, passed: bool, evidence: str = "") -> None:
        self.verification_passed = passed
        self.verification_evidence = evidence
        if passed:
            self.authoritative_artifact = self.final_candidate

    def set_slot(self, node_id: str, content: str) -> None:
        self.slots[node_id] = content

    def get_slot(self, node_id: str) -> str:
        return self.slots.get(node_id, "")

File: src/memory.ts
/**
 * Working memory: shared state for one task (drafts, critique, final artifact, verification result).
 */

/**
 * In-memory scratchpad for a single run of the agent graph.
 */
export class WorkingMemory {
  task = "";
  draft = "";
  critique = "";
  finalCandidate = "";
  verificationPassed = false;
  verificationEvidence = "";
  authoritativeArtifact = "";
  meta: Record<string, unknown> = {};
  /** Graph runner: node_id -> output (filled by graph execution) */
  slots: Record<string, string> = {};
  /** Injected from long-term memory for agent context (read-only) */
  longTermContext = "";

  setTask(task: string): void {
    this.task = task;
  }

  setDraft(draft: string): void {
    this.draft = draft;
  }

  setCritique(critique: string): void {
    this.critique = critique;
  }

  setFinalCandidate(candidate: string): void {
    this.finalCandidate = candidate;
  }

  setVerification(passed: boolean, evidence = ""): void {
    this.verificationPassed = passed;
    this.verificationEvidence = evidence;
    if (passed) {
      this.authoritativeArtifact = this.finalCandidate;
    }
  }

  setSlot(nodeId: string, content: string): void {
    this.slots[nodeId] = content;
  }

  getSlot(nodeId: string): string {
    return this.slots[nodeId] ?? "";
  }
}
